File: src/main/java/ru/ifmo/ctddev/isaev/result/Point.java
package ru.ifmo.ctddev.isaev.result;

import ru.ifmo.ctddev.isaev.melif.impl.FeatureSelectionAlgorithm;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import java.util.stream.Collectors;
import java.util.stream.DoubleStream;
import java.util.stream.IntStream;


/**
 * @author iisaev
 */
public class Point implements Comparable<Point> {

    private static final Consumer<double[]> DO_NOTHING = (d) -> {
    };

    private final double[] coordinates;

    private final int generation;

    public Point(double... coordinates) {
        this(0, coordinates);
    }

    public Point(int generation, double... coordinates) {
        this(generation, DO_NOTHING, coordinates);
    }

    public Point(int generation, Consumer<double[]> consumer, double... coordinates) {
        this.coordinates = coordinates.clone();
        consumer.accept(this.coordinates);
        double modulus = DoubleStream.of(coordinates).sum();
        IntStream.range(0, coordinates.length).forEach(i -> this.coordinates[i] /= modulus); // normalization
        this.generation = generation;
    }

    public Point(Point point) {
        this(point.getCoordinates().clone());
    }

    public Point(Point point, Consumer<double[]> consumer) {
        this.coordinates = point.getCoordinates().clone();
        consumer.accept(coordinates);
        //double modulus = Math.sqrt(DoubleStream.of(coordinates).map(d -> d * d).sum());
        double modulus = DoubleStream.of(coordinates).sum();
        IntStream.range(0, coordinates.length).forEach(i -> this.coordinates[i] /= modulus); // normalization
        this.generation = point.generation + 1;
    }

    public double[] getCoordinates() {
        return coordinates;
    }

    @Override
    public int compareTo(Point other) {
        for (int i = 0; i < coordinates.length; i++) {
            if (coordinates[i] - other.coordinates[i] > 0.001) {
                return 1;
            }
            if (coordinates[i] - other.coordinates[i] < -0.001) {
                return -1;
            }
        }
        return 0;
    }

    @Override
    public String toString() {
        List<String> doubles = DoubleStream.of(coordinates).mapToObj(FeatureSelectionAlgorithm.FORMAT::format).collect(Collectors.toList());
        return "[" + String.join(", ", doubles) + "]/" + (generation != 0 ? generation : "");
    }

    public List<Point> getNeighbours(double delta) {
        List<Point> points = new ArrayList<>();
        IntStream.range(0, coordinates.length).forEach(i -> {
            Point plusDelta = new Point(this, coords -> coords[i] += delta);
            Point minusDelta = new Point(this, coords -> coords[i] -= delta);
            points.add(plusDelta);
            points.add(minusDelta);
        });
        return points;
    }
}

File: melif/src/main/kotlin/ru/ifmo/ctddev/isaev/point/point.kt
package ru.ifmo.ctddev.isaev.point

import ru.ifmo.ctddev.isaev.melif.impl.FeatureSelectionAlgorithm
import java.util.*
import java.util.function.Consumer
import java.util.stream.DoubleStream
import java.util.stream.IntStream

/**
 * @author iisaev
 */
open class Point : Comparable<Point> {
    companion object {
        fun fromRawCoords(vararg coordinates: Double): Point {
            return Point(true, *coordinates)
        }
    }

    val coordinates: DoubleArray

    private val generation: Int

    constructor(vararg coordinates: Double) : this(0, *coordinates)

    constructor(generation: Int,
                vararg coordinates: Double) : this(generation, Consumer {}, *coordinates)

    constructor(generation: Int,
                modifyCoordinates: Consumer<DoubleArray>,
                vararg coordinates: Double) {
        this.coordinates = coordinates.clone()
        modifyCoordinates.accept(this.coordinates)
        val modulus = DoubleStream.of(*coordinates).sum()
        IntStream.range(0, coordinates.size).forEach { i -> this.coordinates[i] /= modulus } // normalization
        this.generation = generation
    }

    constructor(point: Point) : this(*point.coordinates.clone())

    private constructor(unused: Boolean, vararg coordinates: Double) {
        this.coordinates = coordinates
        this.generation = -1
    }

    constructor(point: Point,
                modifyCoordinates: (DoubleArray) -> Unit) {
        this.coordinates = point.coordinates.clone()
        modifyCoordinates(coordinates)
        //double modulus = Math.sqrt(DoubleStream.of(coordinates).map(d -> d * d).sum());
        val modulus = DoubleStream.of(*coordinates).sum()
        IntStream.range(0, coordinates.size).forEach { i -> this.coordinates[i] /= modulus } // normalization
        this.generation = point.generation + 1
    }

    override fun compareTo(other: Point): Int {
        for (i in coordinates.indices) {
            if (coordinates[i] - other.coordinates[i] > 0.00001) {
                return 1
            }
            if (coordinates[i] - other.coordinates[i] < -0.00001) {
                return -1
            }
        }
        return 0
    }

    override fun toString(): String {
        val doubles = coordinates.map { FeatureSelectionAlgorithm.FORMAT.format(it) }
        return "[" + doubles.joinToString(", ") + "]/" + if (generation > 0) generation else ""
    }

    fun getNeighbours(delta: Double): List<Point> {
        val points = ArrayList<Point>()
        IntStream.range(0, coordinates.size - 1).forEach { i ->
            val plusDelta = Point(this, { coords -> coords[i] += delta })
            val minusDelta = Point(this, { coords -> coords[i] -= delta })
            points.add(plusDelta)
            points.add(minusDelta)
        }
        return points
    }

    operator fun get(i: Int): Double = coordinates[i]
}

class PriorityPoint(var priority: Double, p: Point) : Point(p) {

    constructor(value: Point) : this(1.0, value)
}